firmware/f7/Cam/Src/ov5640_driver.c

```diff
@@ -0,0 +1,385 @@
+	uint32_t index = 0;
+	for(index=0; index<(sizeof(format_rgb565)/4); index++)
+	{
+	   CAMERA_IO_Write_OV5640(OV5640_I2C_ADDRESS, format_rgb565[index][0], format_rgb565[index][1]);
+	}
+}
+
+void ov5640_Init_JPEG(uint16_t x_res, uint16_t y_res, uint16_t FifoWidth, uint16_t packetNumber, uint8_t compRatio)
+{
+
+  uint8_t tmp = 0;
+
+//1. set default register values
+  ov5640_Init(); 
+
+
+  //
+
+//2. set pixel format
+  ov5640_Set_JPEG();
+  
+//3. Set output size (pixel format)
+  OV5640_Set_Size(4, 0, x_res, y_res);
+
+//4. configure timings for the output size
+ uint8_t sys_mul = 160;
+ //set_pll(false, sys_mul, 4, 2, false, 2, true, 4);
+
+
+  //set timings
+
+  tmp = CAMERA_IO_Read_OV5640(OV5640_I2C_ADDRESS,OV5640_TIMING_TC_REG21);
+  tmp |= (1 << 5);
+  CAMERA_IO_Write_OV5640(OV5640_I2C_ADDRESS, OV5640_TIMING_TC_REG21, tmp);
+
+  tmp = CAMERA_IO_Read_OV5640(OV5640_I2C_ADDRESS,OV5640_SYSREM_RESET02);
+  tmp &= ~((1 << 4) | (1 << 3) | (1 << 2));
+  CAMERA_IO_Write_OV5640(OV5640_I2C_ADDRESS, OV5640_SYSREM_RESET02, tmp);
+
+
+  tmp = CAMERA_IO_Read_OV5640(OV5640_I2C_ADDRESS,OV5640_CLOCK_ENABLE02);
+  tmp |= ((1 << 5) | (1 << 3));
+  CAMERA_IO_Write_OV5640(OV5640_I2C_ADDRESS, OV5640_CLOCK_ENABLE02, tmp);
+
+  //set polaratis
+
+  tmp = (uint8_t)(OV5640_POLARITY_PCLK_HIGH << 5U) | (OV5640_POLARITY_HREF_HIGH << 1U) | OV5640_POLARITY_VSYNC_HIGH;
+  CAMERA_IO_Write_OV5640(OV5640_I2C_ADDRESS, OV5640_POLARITY_CTRL, tmp);
+
+
+  OV5640_SetPCLK(OV5640_PCLK_23FPS);
+  OV5640_Set_Comp_Ratio(compRatio);
+  OV5640_Config_FIFO(FifoWidth, packetNumber);
+
+    //CAMERA_IO_Write_OV5640(OV5640_I2C_ADDRESS, 0x3035, 0X41); // slow down OV5640 clocks //Turned off --> TURN BACK ON?
+  //CAMERA_IO_Write_OV5640(OV5640_I2C_ADDRESS, 0x3036, 0x79);
+
+
+  
+
+
+}
+
+void set_pll(bool bypass, uint8_t multiplier, uint8_t sys_div, uint8_t pre_div, bool root_2x, uint8_t pclk_root_div, bool pclk_manual, uint8_t pclk_div)
+{
+   //int ret = 0;
+
+   CAMERA_IO_Write_OV5640(OV5640_I2C_ADDRESS, OV5640_SC_PLL_CONTRL5, bypass?0x80:0x00);
+   CAMERA_IO_Write_OV5640(OV5640_I2C_ADDRESS, OV5640_SC_PLL_CONTRL0, 0x1A);
+   CAMERA_IO_Write_OV5640(OV5640_I2C_ADDRESS, OV5640_SC_PLL_CONTRL1, 0x01 | ((sys_div & 0x0f) << 4));
+   CAMERA_IO_Write_OV5640(OV5640_I2C_ADDRESS, OV5640_SC_PLL_CONTRL2, multiplier & 0xff);
+   CAMERA_IO_Write_OV5640(OV5640_I2C_ADDRESS, OV5640_SC_PLL_CONTRL3, (pre_div & 0xf) | (root_2x?0x10:0x00));
+   CAMERA_IO_Write_OV5640(OV5640_I2C_ADDRESS, OV5640_SYSTEM_ROOT_DIVIDER, (pclk_root_div & 0x3) << 4 | 0x06);
+   CAMERA_IO_Write_OV5640(OV5640_I2C_ADDRESS, PCLK_RATIO, pclk_div & 0x1f);
+   CAMERA_IO_Write_OV5640(OV5640_I2C_ADDRESS, VFIFO_CTRL0C, pclk_manual?0x22:0x20);
+   CAMERA_IO_Write_OV5640(OV5640_I2C_ADDRESS, OV5640_SCCB_SYSTEM_CTRL1, 0x13);
+}
+
+
+
+
+int32_t OV5640_SetPCLK(uint32_t ClockValue)
+{
+
+  uint8_t tmp;
+
+  switch(ClockValue)
+  {
+  case OV5640_PCLK_7M:
+    tmp = 0x38;
+    CAMERA_IO_Write_OV5640(OV5640_I2C_ADDRESS, OV5640_SC_PLL_CONTRL2, tmp);
```

Comment:
OV5640 datasheet says the following for OV5640_SC_PLL_CONTRL2:

Bit[3:0]: PLL pre-divider
1, 2, 3, 4, 6, 8

This is a bit confusing to me - I think it might be trying to say that register values map to the divisor as follows:
0x00 -> 1
0x01 -> 2
0x02 -> 3
0x03 -> 4
0x04 -> INVALID
0x05 -> 6
0x06 -> INVALID
0x07 -> 8

i.e. add one to conventional interpretation of the byte, then 5 and 7 are not allowed. Maybe we could add a shim function to go from the divisor to this reg value?

Reply:
Or maybe it is

0x00 -> 1
0x01 -> 2
0x02 -> 3
0x03 -> 4
0x04 -> 6
0x05 -> 8
0x06 -> INVALID
0x07 -> INVALID

Have to test I think